server/src/main/java/org/elasticsearch/action/ActionRunnable.java

```diff
@@ -58,6 +59,35 @@ public String toString() {
         });
     }
 
+    /**
+     * Same as {@link #supply(ActionListener, CheckedSupplier)} but the supplier always returns an object of reference counted result type
+     * which will have its reference count decremented after invoking the listener.
+     * @param listener Listener to invoke
+     * @param supplier Supplier that provides the reference counted value to pass to the listener
+     * @return Wrapped {@code Runnable}
+     */
+    public static <T extends RefCounted> ActionRunnable<T> supplyRefCounted(
```

Comment:
Naming nit/suggestion, how about this?

```suggestion
    public static <T extends RefCounted> ActionRunnable<T> supplyAndDecRef(
```